File: src/main/java/bot/inker/ankh/core/world/storage/DatabaseBackend.java
package bot.inker.ankh.core.world.storage;

import bot.inker.ankh.core.api.storage.ChunkStorage;
import bot.inker.ankh.core.api.world.storage.BlockStorageEntry;
import bot.inker.ankh.core.api.world.storage.StorageBackend;
import bot.inker.ankh.core.common.entity.LocationEmbedded;
import bot.inker.ankh.core.common.entity.WorldChunkEmbedded;
import bot.inker.ankh.core.database.DatabaseService;
import bot.inker.ankh.core.world.entity.StoredBlockEntity;
import lombok.val;

import javax.annotation.Nonnull;
import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.List;
import java.util.stream.Stream;

@Singleton
public class DatabaseBackend implements StorageBackend {
  private final DatabaseService databaseService;

  @Inject
  private DatabaseBackend(DatabaseService databaseService) {
    this.databaseService = databaseService;
  }

  @Override
  public @Nonnull Stream<BlockStorageEntry> provide(@Nonnull ChunkStorage chunkStorage) {
    try (val session = databaseService.getSessionFactory().openSession()) {
      return session.createQuery("select E from stored_block E where E.location.chunk = :chunk", BlockStorageEntry.class)
        .setParameter("chunk", WorldChunkEmbedded.warp(chunkStorage))
        .list()
        .stream();
    }
  }

  @Override
  public void store(@Nonnull ChunkStorage chunkStorage, @Nonnull List<BlockStorageEntry> entries) {
    try (val session = databaseService.getSessionFactory().openSession()) {
      session.beginTransaction();

      session.createQuery("delete from stored_block E where E.location.chunk = :chunk", null)
        .setParameter("chunk", chunkStorage)
        .executeUpdate();
      for (val entry : entries) {
        val entity = new StoredBlockEntity(
          LocationEmbedded.warp(entry.location()),
          entry.blockId(),
          entry.content()
        );
        session.persist(entity);
      }

      session.getTransaction().commit();
    }
  }
}

File: src/main/kotlin/bot/inker/ankh/core/world/storage/DatabaseBackend.kt
package bot.inker.ankh.core.world.storage

import bot.inker.ankh.core.common.entity.WorldChunkEmbedded
import bot.inker.ankh.core.database.DatabaseService
import bot.inker.ankh.core.world.entity.StoredBlockEntity
import java.util.stream.Stream
import javax.inject.Inject

class DatabaseBackend @Inject private constructor(
  private val databaseService: DatabaseService,
) : StorageBackend {
  override fun provide(worldChunk: WorldChunkEmbedded): Stream<BlockStorageEntry> {
    var result: Stream<BlockStorageEntry> = Stream.empty()
    databaseService.sessionFactory.inSession { session ->
      result = session.createQuery(
        "select E from stored_block E where E.location.chunk = :chunk",
        StoredBlockEntity::class.java
      )
        .setParameter("chunk", worldChunk)
        .list()
        .stream()
        .map { it as BlockStorageEntry }
    }
    return result
  }

  override fun store(worldChunk: WorldChunkEmbedded, entries: List<BlockStorageEntry>) {
    databaseService.sessionFactory.inTransaction { session ->
      session.createQuery<Unit>("delete from stored_block E where E.location.chunk = :chunk", null)
        .setParameter("chunk", worldChunk)
        .executeUpdate()
      entries.map {
        StoredBlockEntity().apply {
          this.location = it.location()
          this.blockId = it.blockId().asString()
          this.content = it.content()
        }
      }.forEach(session::persist)
    }
  }
}